File: visual-test-lab/scripts/hexswitch_api.py
#!/usr/bin/env python3
"""
hexSwitch API Server
Provides REST API for Visual Test Lab to interact with real hexSwitch runtime.
"""

from http.server import BaseHTTPRequestHandler, HTTPServer
import json
import logging
from pathlib import Path
import sys
from urllib.parse import urlparse

# Add parent directory to path to import hexswitch
sys.path.insert(0, str(Path(__file__).parent.parent.parent / "src"))

try:
    from hexswitch.config import ConfigError, load_config, validate_config
    from hexswitch.runtime import Runtime, build_execution_plan
except ImportError as e:
    print(f"Error importing hexswitch: {e}", file=sys.stderr)
    print("Make sure hexSwitch is installed: pip install -e .", file=sys.stderr)
    sys.exit(1)

logging.basicConfig(level=logging.INFO)
logger = logging.getLogger(__name__)


class HexSwitchAPIHandler(BaseHTTPRequestHandler):
    """HTTP request handler for hexSwitch API."""

    def do_GET(self):
        """Handle GET requests."""
        parsed_path = urlparse(self.path)
        path = parsed_path.path

        if path == "/api/adapters":
            self._get_adapters()
        elif path == "/api/config":
            self._get_config()
        elif path == "/api/plan":
            self._get_execution_plan()
        else:
            self._send_response(404, {"error": "Not found"})

    def do_POST(self):
        """Handle POST requests."""
        parsed_path = urlparse(self.path)
        path = parsed_path.path

        if path == "/api/config/validate":
            self._validate_config()
        elif path == "/api/config/load":
            self._load_config()
        else:
            self._send_response(404, {"error": "Not found"})

    def _get_adapters(self):
        """Get list of available adapters."""
        # Import from the export script in the same directory
        import importlib.util

        export_script = Path(__file__).parent / "export_adapters.py"
        spec = importlib.util.spec_from_file_location("export_adapters", export_script)
        export_module = importlib.util.module_from_spec(spec)
        spec.loader.exec_module(export_module)

        adapters = export_module.get_adapter_metadata()
        self._send_response(200, {"adapters": adapters})

    def _get_config(self):
        """Get current configuration."""
        config_path = Path(__file__).parent.parent.parent / "hex-config.yaml"
        try:
            config = load_config(config_path)
            self._send_response(200, {"config": config})
        except Exception as e:
            self._send_response(500, {"error": str(e)})

    def _get_execution_plan(self):
        """Get execution plan from current config."""
        config_path = Path(__file__).parent.parent.parent / "hex-config.yaml"
        try:
            config = load_config(config_path)
            plan = build_execution_plan(config)
            self._send_response(200, {"plan": plan})
        except Exception as e:
            self._send_response(500, {"error": str(e)})

    def _validate_config(self):
        """Validate configuration from request body."""
        content_length = int(self.headers.get("Content-Length", 0))
        body = self.rfile.read(content_length)

        try:
            data = json.loads(body.decode("utf-8"))
            config = data.get("config", {})

            validate_config(config)
            self._send_response(200, {"valid": True})
        except ConfigError as e:
            self._send_response(400, {"valid": False, "error": str(e)})
        except Exception as e:
            self._send_response(500, {"error": str(e)})

    def _load_config(self):
        """Load configuration from file path in request body."""
        content_length = int(self.headers.get("Content-Length", 0))
        body = self.rfile.read(content_length)

        try:
            data = json.loads(body.decode("utf-8"))
            config_path = data.get("path", "hex-config.yaml")

            config = load_config(config_path)
            self._send_response(200, {"config": config})
        except Exception as e:
            self._send_response(500, {"error": str(e)})

    def _send_response(self, status_code, data):
        """Send JSON response."""
        self.send_response(status_code)
        self.send_header("Content-Type", "application/json")
        self.send_header("Access-Control-Allow-Origin", "*")
        self.end_headers()

        response = json.dumps(data).encode("utf-8")
        self.wfile.write(response)

    def log_message(self, format, *args):
        """Override to use our logger."""
        logger.debug(f"{self.address_string()} - {format % args}")


def main():
    """Main entry point."""
    port = 8080
    if len(sys.argv) > 1:
        port = int(sys.argv[1])

    server = HTTPServer(("localhost", port), HexSwitchAPIHandler)
    logger.info(f"hexSwitch API server running on http://localhost:{port}")
    logger.info("Endpoints:")
    logger.info("  GET  /api/adapters - List available adapters")
    logger.info("  GET  /api/config - Get current configuration")
    logger.info("  GET  /api/plan - Get execution plan")
    logger.info("  POST /api/config/validate - Validate configuration")
    logger.info("  POST /api/config/load - Load configuration from file")

    try:
        server.serve_forever()
    except KeyboardInterrupt:
        logger.info("Shutting down server...")
        server.shutdown()


if __name__ == "__main__":
    main()

File: visual-test-lab/src/services/hexSwitchAPI.ts
/**
 * hexSwitch API Client
 * 
 * Communicates with the real hexSwitch Python API server
 * to load configurations, validate, and get adapter metadata.
 */

const API_BASE_URL = 'http://localhost:8080';

export interface HexSwitchAdapter {
  id: string;
  name: string;
  type: 'inbound' | 'outbound';
  category: string;
  icon: string;
  description: string;
  source: string;
  configSchema: Record<string, any>;
}

export interface HexSwitchConfig {
  service: {
    name: string;
    runtime?: string;
  };
  inbound?: {
    [adapterName: string]: {
      enabled: boolean;
      [key: string]: any;
    };
  };
  outbound?: {
    [adapterName: string]: {
      enabled: boolean;
      [key: string]: any;
    };
  };
  logging?: {
    level?: 'DEBUG' | 'INFO' | 'WARNING' | 'ERROR';
  };
}

export interface ExecutionPlan {
  service: string;
  inbound_adapters: Array<{
    name: string;
    config: Record<string, any>;
  }>;
  outbound_adapters: Array<{
    name: string;
    config: Record<string, any>;
  }>;
}

/**
 * Check if hexSwitch API server is available
 */
export async function checkAPIAvailable(): Promise<boolean> {
  try {
    const response = await fetch(`${API_BASE_URL}/api/adapters`, {
      method: 'GET',
      signal: AbortSignal.timeout(1000)
    });
    return response.ok;
  } catch (error) {
    return false;
  }
}

/**
 * Get list of available adapters from real hexSwitch
 */
export async function getAdapters(): Promise<HexSwitchAdapter[]> {
  const response = await fetch(`${API_BASE_URL}/api/adapters`);
  if (!response.ok) {
    throw new Error(`Failed to fetch adapters: ${response.statusText}`);
  }
  const data = await response.json();
  return data.adapters;
}

/**
 * Load configuration from hexSwitch
 */
export async function loadConfig(configPath?: string): Promise<HexSwitchConfig> {
  if (configPath) {
    const response = await fetch(`${API_BASE_URL}/api/config/load`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json'
      },
      body: JSON.stringify({ path: configPath })
    });
    if (!response.ok) {
      throw new Error(`Failed to load config: ${response.statusText}`);
    }
    const data = await response.json();
    return data.config;
  } else {
    const response = await fetch(`${API_BASE_URL}/api/config`);
    if (!response.ok) {
      throw new Error(`Failed to get config: ${response.statusText}`);
    }
    const data = await response.json();
    return data.config;
  }
}

/**
 * Validate configuration using real hexSwitch validator
 */
export async function validateConfig(config: HexSwitchConfig): Promise<{
  valid: boolean;
  error?: string;
}> {
  const response = await fetch(`${API_BASE_URL}/api/config/validate`, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json'
    },
    body: JSON.stringify({ config })
  });
  
  const data = await response.json();
  
  if (response.ok) {
    return { valid: data.valid };
  } else {
    return { valid: false, error: data.error || 'Validation failed' };
  }
}

/**
 * Get execution plan from real hexSwitch
 */
export async function getExecutionPlan(): Promise<ExecutionPlan> {
  const response = await fetch(`${API_BASE_URL}/api/plan`);
  if (!response.ok) {
    throw new Error(`Failed to get execution plan: ${response.statusText}`);
  }
  const data = await response.json();
  return data.plan;
}

